src/include/storage/storage_structure/disk_array.h

```diff
@@ -276,5 +366,16 @@ class InMemDiskArrayBuilder : public BaseInMemDiskArray<T> {
     void setNumElementsAndAllocateDiskAPsForBuilding(uint64_t newNumElements);
 };
 
+template<typename U>
+class InMemDiskArrayBuilder : public InMemDiskArrayBuilderInternal {
+public:
+    InMemDiskArrayBuilder(FileHandle& fileHandle, common::page_idx_t headerPageIdx,
+        uint64_t numElements, bool setToZero = false)
+        : InMemDiskArrayBuilderInternal(
+              fileHandle, headerPageIdx, numElements, sizeof(U), setToZero) {}
+
+    inline U& operator[](uint64_t idx) { return *(U*)BaseInMemDiskArray::operator[](idx); }
```

Comment:
This function should be marked as `override`.

Reply:
It doesn't override anything. `BaseInMemDiskArray::operator[]` has a different signature; this is just exposing a typed version of that function. I've changed `BaseInMemDiskArray::operator[]` to protected so that it's not accessible from `InMemDiskArrayBuilder`/`InMemDiskArray` except through the typed versions.

Reply:
ah. got it.

Reply:
Setting it to protected doesn't actually work because I'd initially written it so that `InMemDiskArrayBuilder` inherits from `BaseInMemDiskArray`, but `InMemDiskArray` uses composition (i.e. needs public functions). I've made it so they both use composition (like `BaseDiskArray` too) as that keeps the interfaces cleaner.